File: Java/src/com/yourtion/Pattern03/RegisterClient.java
package com.yourtion.Pattern03;

import java.util.List;

/**
 * Created by Yourtion on 10/10/2016.
 */
public class RegisterClient {
    private static CashRegister cashRegister = new CashRegister(0);
    private static PurchaseInvoker purchaseInvoker = new PurchaseInvoker();

    public void run() {
        Command purchase1 = new Purchase(cashRegister, 100);
        Command purchase2 = new Purchase(cashRegister, 50);

        // Invoke commands, check register total.
        purchaseInvoker.executePurchase(purchase1);
        purchaseInvoker.executePurchase(purchase2);
        System.out.println("After purchases: " + cashRegister.getTotal());

        // Replay purchases
        cashRegister.reset();
        System.out.println("Register reset to 0");
        List<Command> purchases = purchaseInvoker.getPurchaseHistory();
        for (Command purchase : purchases) {
            purchase.execute();
        }
        System.out.println("After replay: " + cashRegister.getTotal());
    }
}

File: Scala/src/com/yourtion/Pattern03/RegisterClient.scala
package com.yourtion.Pattern03

import com.yourtion.Pattern03

/**
  * Created by Yourtion on 10/10/2016.
  */
object RegisterClient {

  def run() = {
    val register = new CashRegister(0)

    val purchaseOne = Register.makePurchase(register, 100)
    val purchaseTwo = Register.makePurchase(register, 50)

    Register.executePurchase(purchaseOne)
    Register.executePurchase(purchaseTwo)

    println("After purchases: " + register.total)

    println("Register reset to 0")
    register.total = 0

    for (purchase <- Register.purchases) {
      purchase.apply()
    }

    println("After replay: " + register.total)

  }
}
